### ISSUE TITLE
Negative batch_limit during calibration on custom dataset

### ISSUE THREAD

**Issue by marialinea** [HuguesTHOMAS/KPConv-PyTorch#118]
Hello!

I'm currently trying to train a custom dataset of outdoor scenes with the following point data structure: `x, y, z, intensity, class`. I've used the code and folder structure for the S3DIS dataset as a template when processing the custom data. I ran into trouble when calling the calibration method in `train_CustomDataset.py`. 

If num_workers=10, the for-loop (in the calibration method) that iterates through the dataloader hangs after 20 iterations. Not sure if this is more of a memory issue than a code problem, so I tried setting num_workers to zero. What happens is that the `self.dataset.batch_limit`variable turns negative, mainly because `b > estim_b`, with `b = len(batch.cloud_inds)` and `estim_b += (b - estim_b) / low_pass_T`. The values of `estim_b` and `low_pass_T` are the same as for the S3DIS dataset.

I'm afraid that it is my lack of understanding of what is going on at a deeper level in your code, especially the potential sampling method, that keeps me from solving the problem. When num_workers=10, the variable `b` never exceeds 2, but when num_workers is zero `b`>1000. Could the problem originate from the Dataset object, and thus my preprocessing of my own data? 

Any insight on the problem(s) would be very much appreciated :)

Cheers!
Maria

**Comment by HuguesTHOMAS** [comment]
Hi Maria, 

First, if you want more details for the potential sampling (and other parts of the code), you can have a look at my PhD thesis, I already mention this in a previous issue: https://github.com/HuguesTHOMAS/KPConv-PyTorch/issues/72#issuecomment-769293888

Try to search for the word "potential" in the previous issues on this GitHub, you will find some info in my answers to the various questions.

Now concerning the calibration method, I pushed a modification recently to make it faster, but at the price that it can be unstable, which could be the reason why your batch_limit turns negative. To make the calibration more stable (but slower), you can decrease the variable:

https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/aee5494f323aa8db523cfdcc7246562d082add00/datasets/S3DIS.py#L1162-L1163

It should be the same order of magnitude as the value of `batch_limit` that you expect to have in your network, which is the approximate total number of points in a batch (all input clouds of the batch combined). Can you tell me the value of the following parameters for your experiment:
- `in_radius`
- `first_subsampling_dl`
- `batch_num`
- Is the density of points in your data more or less constant or varying a lot?

A final comment: it is very strange that the behavior changes when you change the number of threads... In my opinion, this should not happen, and it could mean there is a bug in your Custom Dataset class implementation. What are the modifications that you made from the original S3DIS dataset? If you follow the same folder structure, you should have nearly no modification to make (only the part where you load the intensity instead of the color).

**Comment by marialinea** [comment]
Thanks for a swift reply! I've read through a number of the closed issues, but that one had slipped my attention. 

Based on your comment in #90, the values of the above parameters are
* `first_subsampling_dl `= 0.4 - There are roughly 10 points sampled per square meter, so I chose a value that is a bit higher than the minimum distance between two points. The data comes from a lidar scan and is more or less constant throughout the data
* `in_radius` = `first_subsampling_dl `*50
* `batch_num` = 6 - Unchanged from the original S3DIS code



The modifications I have made so far are the ones mentioned in issue #72:
*  The method that prepares the text files to ply files is changed to fit my data, which is stored in the las file format
*  All of the variables related to the colors, which stores three dimensional data, are now related to the intensity in 1D variables. Meaning that I have made changes in the methods `load_subsampled_clouds`, `potential_item` and `random_item`. 
* `in_features_dim` = 2, and thus I have added an extra elif in the following code segment
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/7fefb6a8d38fd304775199777ad01d9f1546e2ff/datasets/S3DIS.py#L402-L411
the elif statment reads
```python
elif self.config.in_features_dim == 2:
    stacked_features = np.hstack((stacked_features, features[:, :1]))
```

Everything else is kept the same, except the variables named 'S3DISxxxxx', they are changed to 'CustomDataxxxxx'

I've tried increasing `first_subsampling_dl`, but it dosen't seem to have any impact. The same goes for reducing the `expected_N` variable. The output from the terminale reads
```console
Step     1  estim_b = 0.01 batch_limit =   1501
Step     2  estim_b = 2.11 batch_limit = -60958
Step   158  estim_b = 1.23 batch_limit = -51682
Step   309  estim_b = 1.05 batch_limit = -47182
Step   468  estim_b = 1.01 batch_limit = -41827
Step   629  estim_b = 1.00 batch_limit = -35760
Step   788  estim_b = 1.00 batch_limit = -29133
Step   949  estim_b = 1.00 batch_limit = -21779
ERROR: It seems that the calibration have not reached convergence. Here are some plot to understand why:
If you notice unstability, reduce the expected_N value
If convergece is too slow, increase the expected_N value
```
Below are the produced error plots, with the first plot being  of `debug_in` and `debug_out` and the second of `debug_b` and `debug_estim_b`.
1. ![debug_in_out](https://user-images.githubusercontent.com/54434822/128470052-aa351833-53d7-4717-91de-3f7bc7113cc1.png)

2. ![debug_b](https://user-images.githubusercontent.com/54434822/128470222-6c1e72cf-8f58-41ff-9464-432c9f941b1f.png)

While debugging I found that the variable `b` is the one that causes all of the trouble when going from step 1 to step 2, with its value first being 2 than in the next iteration being 3402. I'm having trouble understanding the big jump in numerical value. Maybe I have to take a third and forth look into how I process the las files to ply files.

**Comment by marialinea** [comment]
update: Had to reduce the `expected_N` variable with a magnitude of 1e3, which was a lot more than what I first experimented with. I have managed to start the training of the custom dataset! Thank you so much for the feedback that ultimately solved my problem :)

Best wishes!

**Comment by HuguesTHOMAS** [comment]
Hi @marialinea,

Happy to see that your problem was solved.

So it means in the end that you have less than a few hundred points per input sphere, which is really small. Here are a few comments based on my own experiments.

 - In the case of a sparser dataset like yours, where you end up with very small point clouds, you are free to increase the size of your input (a GPU with more than 10GB should handle batches with more than 10 000 points). That means 3 possibilities:
      * Decrease the `first_subsampling_dl`, the network can handle pretty well data even if the density of the input is smaller than the first grid size, and in that way you are sure not to lose any information in the input details.
      * Increase the `in_radius`: this will have a positive effect if you are classifying very big objects, as the network will have better information about them. Also, this means a faster test speed because you classify larger chunks of data at each iteration.
      * Increase the `batch_num`: larger batch can help accelerate the convergence but in my experiments, I did not see much improvement for values above 10 or 16 depending on the task.
 - If you still keep very small input clouds, and modify the `batch_num`, it would be worth playing a little bit with the learning rate decay for faster training. Here:
https://github.com/HuguesTHOMAS/KPConv-PyTorch/blob/cf1f06381ef062344b68286e3d5034b1aa94aecd/train_S3DIS.py#L173

The value `150` means we divide the learning rate by 10 every 150 epochs, but in a smooth way, a little bit at each epoch. You can reduce this value for faster convergence, but when you want to get your final result, set it higher up again to be sure you reach the minimum possible loss.

**Comment by marialinea** [comment]
Thanks for the tips, they'll come in handy!

After some 20 epochs it seems that the network isn't learning at all, a consistent mIoU of 2.1% or less, so I'll have to do some experimenting.

**Comment by aldinorizaldy** [comment]
Hello @marialinea . I had the same problem. The `batch_limit is` negative. But it was solved by reducing the `expected_N` value into 1e4.

Regarding the consisten low mIoU, did you solve your problem? If so, how did you solve it?

**Comment by marialinea** [comment]
Hi @aldinorizaldy, I don't think I ever solved the problem before my summer internship ended. So, unfortunately, I don't have a solution for you, best of luck!

**Comment by aldinorizaldy** [comment]
Hi @marialinea, I solved the problem by changing the `first_subsampling_dl` and `in_radius`.  Apparently, my data is too sparse so I need much bigger values for them. Thanks for responding anyway.
